### PR TITLE
Feat/105 develop the data steps of yolov8 classif

### LINKED ISSUE #105: Develop the Data Steps of Yolov8-classif
**As a DS**
**I want to create the first 3 steps of the pipeline**
**So that I can integrate the Yolov8 with the new repository architecture.**

**Acceptance Criteria:**
- [x] Step Data Extraction is developed
- [x] Step Data Validation is developed
- [x] Step Data Preparation is developed

**Definition of Done:**
- [x] All acceptance criteria are met.
- [ ] Code is reviewed and approved.
- [x] The code is covered by tests.
- [ ] Steps are well documented.

Why 13 SP? First steps developed, write the first tests, Data Validation does not exist yet

### PR DESCRIPTION
**Pull request by SoniaGrh**
## Description
 Creation of the first 3 data steps of the pipeline

## Issues
This PR resolves #105 

## What's new?
-  Step Data Extraction is developed
-  Step Data Validation is developed
-  Step Data Preparation is developed
-  Tests for all new steps with 100% coverage 
image

### CODE REVIEW

File: src/steps/data_preparation/utils/classification_dataset_organizer.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    def _create_category_dir_and_copy_image(self, category_name: str, image: Image):
+        """
+        Creates a directory for a category if it doesn't exist and copies an image into it.
+
+        Parameters:
+            category_name (str): The name of the category.
+            image: The image object containing the file name.
+        """
+        category_dir = os.path.join(self.destination_dir, category_name)
```
Comment (picsalex) [inline src/steps/data_preparation/utils/classification_dataset_organizer.py:95]: This code can raise several exceptions (especially with `os.makedirs` and `shutil.copy`). Maybe add inside the docstring something like:

```python
"""
Raises:
        PermissionError: If there is a permission issue creating the directory or copying the file.
        FileNotFoundError: If the source image file to copy is not found.
        shutil.SameFileError: If the source and destination paths are the same.
"""
```

File: src/steps/data_validation/data_validator.py
Comment (picsalex) [inline src/steps/data_validation/data_validator.py:1]: Is this file necessary? Is there a use case where we might validate a dataset that is neither a classification, detection nor segmentation dataset ?

File: src/steps/data_validation/utils/classification_dataset_validator.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+from src.steps.data_validation.utils.dataset_validator import DatasetValidator
+
+
+class ClassificationDatasetValidator(DatasetValidator):
+    def validate(self):
+        super().validate()  # Call common validations
+        self.validate_labelmap()
+        self.validate_at_least_one_image_per_class()
+
+    def validate_labelmap(self):
```
Comment (picsalex) [inline src/steps/data_validation/utils/classification_dataset_validator.py:12]: Missing docstring

File: src/steps/data_validation/utils/classification_dataset_validator.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        for dataset_context in self.dataset_collection:
+            if len(dataset_context.labelmap) < 2:
+                raise ValueError(
+                    f"Labelmap must have at least 2 classes in {dataset_context.dataset_name} dataset"
```
Comment (picsalex) [inline src/steps/data_validation/utils/classification_dataset_validator.py:16]: I would rephrase to something like `Labelmap for dataset {dataset_context.name} is not valid. A classification labelmap must have at least 2 classes. Current labelmap is {dataset_context.labelmap}`

File: src/steps/data_validation/utils/classification_dataset_validator.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+                )
+
+    def validate_at_least_one_image_per_class(self):
```
Comment (picsalex) [inline src/steps/data_validation/utils/classification_dataset_validator.py:19]: Missing docstring, also why do we perform this check only on the train dataset? Shouldn't we validate the `val` and `test` splits as well?

File: src/steps/data_validation/utils/dataset_validator.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
+import os
+
+from PIL import Image
+
+from src.models.dataset.dataset_collection import DatasetCollection
+from src.models.dataset.dataset_context import DatasetContext
+
+
+def get_image_path_list(image_dir: str) -> list[str]:
```
Comment (picsalex) [inline src/steps/data_validation/utils/dataset_validator.py:9]: The return type should be `List[str]`, otherwise we would lose backward compatibility with Python 3.8 and 3.9

File: src/steps/data_validation/utils/dataset_validator.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        ValueError: If any of the images are found to be corrupted.
+    """
+    for image_path in image_path_list:
+        try:
+            with Image.open(image_path) as img:
+                img.verify()  # Verify that this is a valid image
+        except Exception as e:
+            raise ValueError(
+                f"Image {image_path} is corrupted in {dataset_context.dataset_name} dataset - {e}"
+            )
```
Comment (picsalex) [inline src/steps/data_validation/utils/dataset_validator.py:69]: can be rewritten with

```python
try:
   with Image.open(image_path) as img:
      img.verify()  # Verify that this is a valid image

   except Exception as e:
      raise ValueError(f"Image {image_path} is corrupted in {dataset_context.dataset_name} dataset and cannot be used.") from e

```

File: src/steps/data_validation/utils/dataset_validator.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+        Args:
+            dataset_context (DatasetContext): The dataset context to validate.
+            image_path_list (list[str]): The list of image paths extracted from the dataset.
+
+        Raises:
+            ValueError: If any image is not in one of the valid formats.
+        """
+        for image_path in image_path_list:
+            if not image_path.endswith(self.VALID_IMAGE_EXTENSIONS):
```
Comment (picsalex) [inline src/steps/data_validation/utils/dataset_validator.py:124]: you can apply a `lower()` function since the `VALID_IMAGE_EXTENSIONS` also exist in uppercase. Something like:

```python
(...)
if not image_path.lower().endswith(self.VALID_IMAGE_EXTENSIONS):
(...)
```

File: src/steps/data_validation/utils/dataset_validator.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    def __init__(self, dataset_collection: DatasetCollection):
+        """
+        Initializes the DatasetValidator with a dataset collection to validate.
+
+        Parameters:
+            dataset_collection (DatasetCollection): The dataset collection to validate.
+        """
+        self.dataset_collection = dataset_collection
+
+    def validate_common(self) -> None:
```
Comment (picsalex) [inline src/steps/data_validation/utils/dataset_validator.py:95]: Since this is not the overwritten function (but `def validate()` instead), maybe this one can be prefixed with an underscore `_`?

File: src/steps/data_validation/utils/dataset_validator.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        """
+        Performs common validation tasks across all datasets in the collection.
+        """
+        for dataset_context in self.dataset_collection:
+            image_path_list = get_image_path_list(image_dir=dataset_context.image_dir)
+            validate_image_extraction(
+                dataset_context=dataset_context, image_path_list=image_path_list
```
Comment (picsalex) [inline src/steps/data_validation/utils/dataset_validator.py:102]: some `self` calls are missing

File: src/steps/data_validation/utils/dataset_validator.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+            dataset_context (DatasetContext): The dataset context to validate.
+            image_path_list (list[str]): The list of image paths extracted from the dataset.
+
+        Raises:
+            ValueError: If any image is not in one of the valid formats.
+        """
+        for image_path in image_path_list:
+            if not image_path.endswith(self.VALID_IMAGE_EXTENSIONS):
+                raise ValueError(
+                    f"Invalid image format for image {image_path} in {dataset_context.dataset_name} dataset"
```
Comment (picsalex) [inline src/steps/data_validation/utils/dataset_validator.py:126]: We can add more details, something like `Invalid image format for image {image_path} in {dataset_context.dataset_name} dataset. Valid image formats are {self.VALID_IMAGE_EXTENSIONS}`

File: src/steps/data_validation/utils/dataset_validator.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        try:
+            with Image.open(image_path) as img:
+                img.verify()  # Verify that this is a valid image
+        except Exception as e:
+            raise ValueError(
+                f"Image {image_path} is corrupted in {dataset_context.dataset_name} dataset - {e}"
+            )
+
+
+class DatasetValidator:
```
Comment (picsalex) [inline src/steps/data_validation/utils/dataset_validator.py:72]: is the abstract `ABC` statement missing?

File: src/steps/data_validation/utils/dataset_validator.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+        Raises:
+            ValueError: If any image is not in one of the valid formats.
+        """
+        for image_path in image_path_list:
+            if not image_path.endswith(self.VALID_IMAGE_EXTENSIONS):
+                raise ValueError(
+                    f"Invalid image format for image {image_path} in {dataset_context.dataset_name} dataset"
+                )
+
```
Comment (picsalex) [inline src/steps/data_validation/utils/dataset_validator.py:128]: missing the `@abstractmethod` decorator

File: src/models/parameters/hyper_parameters.py [hunk 13]
```diff
@@ -28,6 +28,10 @@ def __init__(self, log_data: LogDataType):
             default=False,
         )
 
+        self.prop_train_split = self.extract_parameter(
```
[the remark this answers is not in the document]
Comment (SoniaGrh) [reply src/models/parameters/hyper_parameters.py:31]: ChatGPT suggests : 
- adaptive_split_ratio
- auto_split_proportion
- adaptive_train_split
- adaptive_training_proportion

What do you think ?

[the remark this answers is not in the document]
Comment (SoniaGrh) [reply src/models/parameters/hyper_parameters.py:31]: ChatGPT suggests : 
- adaptive_split_ratio
- auto_split_proportion
- adaptive_train_split
- adaptive_training_proportion

What do you think ?

[the remark this answers is not in the document]
Comment (picsalex) [reply src/models/parameters/hyper_parameters.py:31]: what about something like `train_set_split_ratio` ?

File: src/steps/data_validation/utils/classification_dataset_validator.py (hunk 4 above)
Comment (SoniaGrh) [reply src/steps/data_validation/utils/classification_dataset_validator.py:19]: If the validation and test datasets are not well-represented across all classes, it's not necessarily our responsibility to ensure they are, or at least it shouldn't be a blocking issue. We can maybe implement these checks as warnings. Do you think it should be blocking ?

File: src/steps/data_validation/data_validator.py
Comment (SoniaGrh) [reply src/steps/data_validation/data_validator.py:1]: Yes, this file could be useful beyond the scope of classification, detection, or segmentation datasets. For instance, in data preprocessing tasks, which might not directly involve model training, validation can still play a crucial role. It ensures that the data meets certain standards or follows specific formats before it's used. There could be use cases where someone wants to implement their own custom validation logic on top of some basic validations provided by this file.

File: src/steps/data_validation/utils/classification_dataset_validator.py (hunk 4 above)
Comment (picsalex) [reply src/steps/data_validation/utils/classification_dataset_validator.py:19]: Hum, I guess this should be blocking :
- What if, as a user, I did a mistake and did not configured my datasets correctly ?
- What if the split has created an empty dataset split (even if almost impossible) ?

File: src/steps/data_validation/data_validator.py
Comment (picsalex) [reply src/steps/data_validation/data_validator.py:1]: ok 👍

File: src/steps/data_validation/utils/dataset_validator.py (hunk 9 above)
Comment (SoniaGrh) [reply src/steps/data_validation/utils/dataset_validator.py:102]: The functions **get_image_path_list** and **validate_image_extraction** do not depend on the class instance, which is why they don't utilize self. Do you think it's better to include them in this class or are you talking about something else?

Comment (picsalex) [reply src/steps/data_validation/utils/dataset_validator.py:104]: i guess they could be added inside the DatasetValidator class, since they are also prefixed by "validate"

File: src/steps/data_extraction/utils/dataset_handler.py [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+            - DatasetCollection: A collection of dataset contexts prepared based on the attached dataset versions.
+        """
+        nb_attached_datasets = len(self.experiment.list_attached_dataset_versions())
+        try:
+            train_dataset_version = self.experiment.get_dataset(
+                DatasetSplitName.TRAIN.value
+            )
+        except Exception as e:
+            raise ResourceNotFoundError(
+                "Training dataset not found in the experiment."
```
Comment (picsalex) [inline src/steps/data_extraction/utils/dataset_handler.py:75]: Maybe we can improve this message, something like "A training dataset is required and couldn't be found. Please attach a dataset with the alias 'train'."

File: src/steps/data_extraction/utils/dataset_handler.py [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+            ) from e
+
+        if nb_attached_datasets == 3:
+            try:
+                val_dataset_version = self.experiment.get_dataset(
+                    DatasetSplitName.VAL.value
+                )
+            except Exception as e:
+                raise ResourceNotFoundError(
+                    f"Found {nb_attached_datasets} attached datasets but {DatasetSplitName.VAL.value} not found."
```
Comment (picsalex) [inline src/steps/data_extraction/utils/dataset_handler.py:85]:
```python
f"Found {nb_attached_datasets} attached datasets but a validation dataset couldn't be found. Please attach a dataset with the alias 'val'"
```

File: src/steps/data_extraction/utils/dataset_handler.py [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+                ) from e
+            try:
+                test_dataset_version = self.experiment.get_dataset(
+                    DatasetSplitName.TEST.value
+                )
+            except Exception as e:
+                raise ResourceNotFoundError(
+                    f"Found {nb_attached_datasets} attached datasets but {DatasetSplitName.TEST.value} not found"
```
Comment (picsalex) [inline src/steps/data_extraction/utils/dataset_handler.py:93]: same

File: src/steps/data_extraction/utils/dataset_handler.py [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+                test_dataset_version=test_dataset_version,
+            )
+        elif nb_attached_datasets == 2:
+            try:
+                test_dataset_version = self.experiment.get_dataset(
+                    DatasetSplitName.TEST.value
+                )
+            except Exception as e:
+                raise ResourceNotFoundError(
+                    f"Found {nb_attached_datasets} attached datasets but {DatasetSplitName.TEST.value} not found"
```
Comment (picsalex) [inline src/steps/data_extraction/utils/dataset_handler.py:107]: same

**Review by picsalex — APPROVED** [review]
